File: C++/task#1.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

// Алгоритм Евклида(поиск НОД)
int NOD(int a, int p) {
    while (p != 0) {
        int r = a % p;
        a = p;
        p = r;
    }

    return a;
}

// Малая теорема Ферма
int mTeoremaFerma(int a, int p) {
    if (NOD(a, p) != 1) {  // проверка на взаимную простоту 
        return -1;
    }
    int s = 1;
    for (int i = 0; i < p - 1; i++) { // a^p-1 mod p
        s = (s * a) % p; // a*(a mod p) = a*a mod p
    }
    if (s == 1) {
        return 1; // p - вероятно простое
    }
    else {
        return 0; // р - составное
    }
}

// Свойство сравнения 
int svoSrav(int a, int x, int p) {
    int left = alg_2(a, x, p); // (a^x) mod p
    int a_mod = a % p;
    int right = alg_2(a_mod, x, p); // ((a mod p)^x) mod p
    return (left == right) ? 1 : -1;
}

// Модульное умножение
int alg_1(int a, int x, int p) {
    int res = 1;
    for (int i = 1; i <= x; i++) {
        res = res * a; // a*(a mod p) = (a*a) mod p
        res = res % p; // сразу берем остаток, избегая переполнения
    }
    return res;
}

// Бинарное возведение в степень
int alg_2(int a, int x, int p) {
    int res = 1;
    a = a % p;

    while (x > 0) {
        if (x % 2 == 1) {
            res = (res * a) % p;
        }
        a = (a * a) % p;
        x = x / 2;
    }

    return res;
}

int main() {
    setlocale(LC_ALL, "rus");
    int a, x, p;

    cout << "Введите числа a,x,p через пробел >>>> ";
    cin >> a >> x >> p;

    int r1 = mTeoremaFerma(a, p); // проверка р на простоту
    if (r1 == 1) {
        cout << "a^(p-1) mod p = 1, следовательно, число p - вероятно простое!" << endl;
    }
    else if (r1 == -1) {
        cout << "Введенные числа не взаимно простые!" << endl;
        return -1;
    }
    else {
        cout << "a^x mod p != 1, следовательно, число p - СОСТАВНОЕ!" << endl;
        return -1;
    }

    int r2 = svoSrav(a, x, p); // проверка свойства сравнения 
    if (r2 != 1) {
        cout << "Свойство сравнений НЕ выполняется" << endl;
    }

    int alg1 = alg_1(a, x, p);
    cout << "\nАлгоритм №1 - По правилам модульного умножения: a^x mod p = " << alg1 << endl;
    int alg2 = alg_2(a, x, p);
    cout << "Алгоритм №2 - Метод бинарного возведения в степень: a^x mod p = " << alg2 << endl;
    if (alg1 == alg2) {
        cout << "\nДанные совпадают >>>> Алгоритмы работают правильно!" << endl;
    }

    return 0;
}

File: Python/task#1.py
def NOD(a, p):
    while p != 0:
        r = a % p
        a = p
        p = r
    return a

def mTeoremaFerma(a, p):
    if NOD(a, p) != 1:
        return -1
    s = 1
    for _ in range(p - 1):
        s = (s * a) % p
    if s == 1:
        return 1
    else:
        return 0

def svo_srav(a, x, p):
    left = alg_2(a, x, p) # (a^x) mod p
    a_mod = a % p
    right = alg_2(a_mod, x, p) # ((a mod p)^x) mod p
    return 1 if left == right else -1

def alg_1(a, x, p):
    res = 1
    for _ in range(x):
        res = (res * a) % p
    return res

def alg_2(a, x, p):
    res = 1
    a = a % p
    while x > 0:
        if x % 2 == 1:
            res = (res * a) % p
        a = (a * a) % p
        x //= 2
    return res

def main():
    a, x, p = map(int, input("Введите числа a, x, p через пробел >>>> ").split())

    r1 = mTeoremaFerma(a, p)
    if r1 == 1:
        print("a^(p-1) mod p = 1, следовательно, число p - вероятно простое!")
    elif r1 == -1:
        print("Введенные числа не взаимно простые!")
        return
    else:
        print("a^x mod p != 1, следовательно, число p - СОСТАВНОЕ!")
        return

    r2 = svoSrav(a, x, p)
    if r2 != 1:
        print("Свойство сравнений НЕ выполняется")

    alg1 = alg_1(a, x, p)
    print(f"\nАлгоритм №1 - По правилам модульного умножения: a^x mod p = {alg1}")

    alg2 = alg_2(a, x, p)
    print(f"Алгоритм №2 - Метод бинарного возведения в степень: a^x mod p = {alg2}")

    if alg1 == alg2:
        print("\nДанные совпадают >>>> Алгоритмы работают правильно!")

if __name__ == "__main__":
    main()